contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,368 @@ const (
+3. User creates PVC. Nothing changes in the spec, although now the PVC won't be
+   immediately bound.
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: my-pvc
+spec:
+  storageClassName: standard
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 100Gi
+```
+4. User creates Pod. Nothing changes in the spec.
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-pod
+spec:
+  containers:
+  ...
+  volumes:
+  - name: my-vol
+    persistentVolumeClaim:
+      claimName: my-pvc
+```
+5. Scheduler picks a node that can satisfy the Pod and passes it to the
+   provisioner.
+6. Provisioner dynamically provisions a PV that can be accessed from
+   that node.
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: standard
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+```
+7. Pod gets scheduled to the node.
+
+
+### Restricting Topology
+For the common use case, volumes will be provisioned in whatever topology domain
+the scheduler has decided is best to run the workload. Users may impose further
+restrictions by setting label/node selectors, and pod affinity/anti-affinity
+policies on their Pods. All those policies will be taken into account when
+dynamically provisioning a volume.
+
+While less common, administrators may want to further restrict what topology
+domains are available to a StorageClass. To support these administrator
+policies, an AllowedTopology field can also be specified in the
+StorageClass to restrict the topology domains for dynamic provisioning.
+This is not expected to be a common use case, and there are some caveats,
+described below.
+
+```
+type StorageClass struct {
+    ...
+
+    AllowedTopologies *VolumeProvisioningTopologies
+}
+
+type VolumeProvisioningTopologies struct {
+    // Volume must be provisioned from one of the required topologies
+    RequiredTopologies []VolumeTopology
+}
+
+type VolumeTopology struct{
+    // The provisioned volume must support all the specified node labels
+    MatchNodeLabels map[string]string
+}
+```
+
+A nil value means there are no topology restrictions. A scheduler predicate
+will evaluate a non-nil value when considering dynamic provisioning for a node.
+When evaluating the allowed topology against a node, the node needs to have
+labels matching everything specified in a VolumeTopology for one of the
+RequiredTopologies.
+
+A StorageClass update controller can be provided to convert existing topology
+restrictions specified under plugin-specific parameters to the new AllowedTopologies
+field.
+
+The AllowedTopologies will also be provided to provisioners as a new field, detailed in
```

Comment:
It would be nice to have a sentence or two why node selector was not used in StorageClass.

Reply:
added